src/pretty/markup.rs

```diff
@@ -47,10 +132,116 @@ impl<'a> PrettyPrinter<'a> {
+                // Here we can safely strip the space. Should never turn space into line.
+                let doc = self.convert_markup_stripped(markup);
+                FlowItem::spaced(doc.nest(2))
             } else {
                 FlowItem::none()
             }
         })
     }
 }
+
+struct MarkupItem<'a> {
+    node: &'a SyntaxNode,
+    format_disabled: bool,
+}
+
+struct MarkupItems<'a> {
+    items: Vec<MarkupItem<'a>>,
+    has_leading_space: bool,
+    has_trailing_space: bool,
+}
+
+// Break markup into lines, split by stmt, parbreak, newline, multiline raw,
+// equation if a line contains text, it will be skipped by the formatter
+// to keep the original format.
+fn collect_markup_items(markup: Markup<'_>) -> MarkupItems {
+    let mut items = MarkupItems {
+        items: vec![],
+        has_leading_space: false,
+        has_trailing_space: false,
+    };
+    let mut cursor = 0;
+    let mut current_line_has_text = false;
+    for node in markup.to_untyped().children() {
+        let mut break_line = false;
+        if (node.kind() == SyntaxKind::Space || node.kind() == SyntaxKind::Parbreak)
+            && node.text().has_linebreak()
+            || node.kind().is_stmt()
+        {
+            break_line = true;
+        } else if let Some(expr) = node.cast::<Expr>() {
+            match expr {
+                Expr::Text(_) => current_line_has_text = true,
+                Expr::Raw(r) => {
+                    if r.block() {
+                        break_line = true;
+                    } else {
+                        current_line_has_text = true;
+                    }
+                }
+                Expr::Strong(_) | Expr::Emph(_) => current_line_has_text = true,
+                Expr::Code(_) => break_line = true,
+                Expr::Equation(e) if e.block() => break_line = true,
+                _ => (),
+            }
+        }
+        if node.kind() == SyntaxKind::Space && items.items.is_empty() {
+            // Discard leading space and mark it.
+            items.has_leading_space = true;
+        } else {
+            items.items.push(MarkupItem {
+                node,
+                format_disabled: false,
+            });
+        }
+        if break_line {
+            if current_line_has_text {
+                for item in &mut items.items[cursor..] {
+                    item.format_disabled = true;
+                }
+            }
+            cursor = items.items.len();
+            current_line_has_text = false;
+        }
+    }
+    if current_line_has_text {
+        for item in &mut items.items[cursor..] {
+            item.format_disabled = true;
+        }
+    }
+
+    // Remove trailing spaces
+    while items
+        .items
+        .last()
+        .is_some_and(|last| last.node.kind() == SyntaxKind::Space)
+    {
+        items.items.pop();
+        items.has_trailing_space = true;
+    }
+
+    // Check space inside comments
+    if !items.has_leading_space
```

Comment:
is it possible to fuse comment related check into previous loop? i.e. we can loop through each children and ignore comment nodes.